grin/test/Transformations/Optimising/CSESpec.hs

```diff
@@ -15,10 +18,16 @@ import Transformations.EffectMap
 runTests :: IO ()
 runTests = hspec spec
 
-cseOptNoEff (tyEnv, exp) = commonSubExpressionElimination tyEnv mempty exp
+pipelineSrcWithCtx :: ((TypeEnv, Exp) -> (TypeEnv, Exp)) ->
+                      TypeEnv -> Exp -> Exp ->
+                      [PipelineStep] -> Pipeline
+pipelineSrcWithCtx ctx tyEnv before after =
+  pipelineSrc (snd $ ctx (tyEnv, before)) (snd $ ctx (tyEnv, after))
```

Comment:
I think this extra indirection makes the readability worse. Also I am not sure anymore that the ctx idea worthwhile.

But definetly we need to run a cleanup round around the first release, until than I would postpone any non-trivial cleanup.